File: WKUAdventure/Game/renpy.ts
export type RoleId = "narrator" | "lin-cheng" | "jiang-ran" | "tang-li" | "zhou-yuan";
export type StatKey = "chaos" | "order" | "warmth" | "wit";

export interface Effect {
	key: StatKey;
	amount: number;
}

export interface MenuChoice {
	id: string;
	text: string;
	effects: Effect[];
	jump: string;
}

export type Statement =
	| {kind: "label"; name: string}
	| {kind: "say"; speaker: RoleId; text: string}
	| {kind: "signal"; signalId: string}
	| {kind: "menu"; prompt: string; choices: [MenuChoice, MenuChoice]}
	| {kind: "jump"; target: string}
	| {kind: "condition"; key: StatKey; minimum: number; pass: string; fail: string}
	| {kind: "minigame"; gameId: string; fallback: string}
	| {kind: "end"; id: string; title: string; summary: string; reveal?: string};

export interface StoryStats {
	chaos: number;
	order: number;
	warmth: number;
	wit: number;
}

export class RenpyRuntime {
	private cursor = 0;
	private readonly labels: Record<string, number> = {};

	constructor(private readonly script: Statement[], private readonly stats: StoryStats) {
		for (let index = 0; index < script.length; index++) {
			const statement = script[index];
			if (statement.kind === "label") this.labels[statement.name] = index + 1;
		}
	}

	public start(label = "start"): Statement | undefined {
		this.jump(label);
		return this.nextVisible();
	}

	public resume(cursor: number): Statement | undefined {
		this.cursor = math.max(0, math.min(cursor, this.script.length));
		return this.nextVisible();
	}

	public advance(): Statement | undefined {
		return this.nextVisible();
	}

	public choose(choice: MenuChoice): Statement | undefined {
		for (const effect of choice.effects) this.stats[effect.key] += effect.amount;
		this.jump(choice.jump);
		return this.nextVisible();
	}

	public completeMiniGame(statement: Extract<Statement, {kind: "minigame"}>): Statement | undefined {
		this.jump(statement.fallback);
		return this.nextVisible();
	}

	public getCursor(): number {
		return this.cursor;
	}

	private jump(label: string): void {
		const target = this.labels[label];
		if (target === undefined) error(`剧情标签不存在: ${label}`);
		this.cursor = target;
	}

	private nextVisible(): Statement | undefined {
		while (this.cursor < this.script.length) {
			const statement = this.script[this.cursor];
			this.cursor += 1;
			if (statement.kind === "label") continue;
			if (statement.kind === "jump") {
				this.jump(statement.target);
				continue;
			}
			if (statement.kind === "condition") {
				this.jump(this.stats[statement.key] >= statement.minimum ? statement.pass : statement.fail);
				continue;
			}
			return statement;
		}
		return undefined;
	}
}

File: WKUAdventure/Game/renpy.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local ____exports = {}
____exports.RenpyRuntime = __TS__Class()
local RenpyRuntime = ____exports.RenpyRuntime
RenpyRuntime.name = "RenpyRuntime"
function RenpyRuntime.prototype.____constructor(self, script, stats)
    self.script = script
    self.stats = stats
    self.cursor = 0
    self.labels = {}
    do
        local index = 0
        while index < #script do
            local statement = script[index + 1]
            if statement.kind == "label" then
                self.labels[statement.name] = index + 1
            end
            index = index + 1
        end
    end
end
function RenpyRuntime.prototype.start(self, label)
    if label == nil then
        label = "start"
    end
    self:jump(label)
    return self:nextVisible()
end
function RenpyRuntime.prototype.resume(self, cursor)
    self.cursor = math.max(
        0,
        math.min(cursor, #self.script)
    )
    return self:nextVisible()
end
function RenpyRuntime.prototype.advance(self)
    return self:nextVisible()
end
function RenpyRuntime.prototype.choose(self, choice)
    for ____, effect in ipairs(choice.effects) do
        local ____self_stats_0, ____effect_key_1 = self.stats, effect.key
        ____self_stats_0[____effect_key_1] = ____self_stats_0[____effect_key_1] + effect.amount
    end
    self:jump(choice.jump)
    return self:nextVisible()
end
function RenpyRuntime.prototype.completeMiniGame(self, statement)
    self:jump(statement.fallback)
    return self:nextVisible()
end
function RenpyRuntime.prototype.getCursor(self)
    return self.cursor
end
function RenpyRuntime.prototype.jump(self, label)
    local target = self.labels[label]
    if target == nil then
        error("剧情标签不存在: " .. label)
    end
    self.cursor = target
end
function RenpyRuntime.prototype.nextVisible(self)
    while self.cursor < #self.script do
        do
            local __continue17
            repeat
                local statement = self.script[self.cursor + 1]
                self.cursor = self.cursor + 1
                if statement.kind == "label" then
                    __continue17 = true
                    break
                end
                if statement.kind == "jump" then
                    self:jump(statement.target)
                    __continue17 = true
                    break
                end
                if statement.kind == "condition" then
                    self:jump(self.stats[statement.key] >= statement.minimum and statement.pass or statement.fail)
                    __continue17 = true
                    break
                end
                return statement
            until true
            if not __continue17 then
                break
            end
        end
    end
    return nil
end
return ____exports
